### PR TITLE
feat: docs i18n support (continuation from #14818)

### PR DESCRIPTION
**Pull request by timhagn**
**Problem**

Our documentation is only written in English (and Chinese). This PR attempts to add support via Docusaurus' internationalization features.

Summary of Changes:

- [x] add i18n for the docs
- [x] merge in latest master
- [x] add Crowdin support & `crowdin:sync` command for CI
- [x] add languages setup in Crowdin & add German (de)
- [x] add failing locales in `docusaurus.config.js` as `// localesNotBuildingLocally: ["ko", "pt", "vi"],` comment
- [x] add `CROWDIN_PERSONAL_TOKEN` to Travis env variables
- [x] add call to `npm run crowdin:download` to Travis script

**Still TODO**

- [ ] look into failing locales
- [ ] add `CROWDIN_PERSONAL_TOKEN` to Vercel env variables
- [ ] Localize homepage (use json api)

### CODE REVIEW

**Comment by jstarry** [comment]
Awesome @timhagn! Here's some high level feedback:

1. Keep translations out of the repo

Yes, I started this in https://github.com/solana-labs/solana/pull/17125 but I realized since then that it would be much better to download the translations as part of the doc CI build rather than include them in source. It only makes sense to keep them in source if we try to sync github changes back to crowdin, which docusaurus recommends not doing.

2. Add explicit headings

Since headings are translated, we need explicit headings to keep the slug the same and avoid breaking links. I don't think we need to include these explicit headings in source control so the best route is to run heading generation right before english source files are uploaded in CI. Info: https://docusaurus.io/docs/cli#docusaurus-write-heading-ids-sitedir

**Comment by jstarry** [comment]
Versioning is hard so I think the best workflow for now is automatically uploading translations after each commit in the stable branch (currently v1.6) and setting up automatic PR's using Crowdin's github integration to sync translations back to the stable branch. This workflow means the translations will always be delayed and we won't have translations for beta and edge but I think that's ok for v1

**Comment by timhagn** [comment]
Hi @jstarry,

1. removed the `i18n` folder from git's cache
2. added a `write-heading-ids` script for it

Pertaining the suggested process for versioning the docs, I just added `crowdin:sync` to the `before_install.sh` script at first,
but you're probably right, upload should only be done manually or for the `master` branch.
So I added a `crowdin:download` script & running only that in `before_install.sh`, or would you also want to remove that?

Best,

Tim.

File: docs/package.json [hunk 1]
```diff
@@ -9,12 +9,20 @@
...
+    "write-i18n": "npm run write-heading-ids && npm run write-translations",
     "format": "prettier --check \"**/*.{js,jsx,json,md,scss}\"",
     "format:fix": "prettier --write \"**/*.{js,jsx,json,md,scss}\"",
     "lint": "set -ex; eslint .",
-    "lint:fix": "npm run lint -- --fix"
+    "lint:fix": "npm run lint -- --fix",
+    "crowdin": "crowdin",
+    "crowdin:download": "npm run write-i18n && npm run crowdin download && npm run format:fix",
+    "crowdin:sync": "npm run write-i18n && npm run crowdin upload && npm run crowdin download && npm run format:fix",
+    "prebuild": "npm run crowdin:download"
```
Comment (jstarry) [inline docs/package.json:22]: I would like to be able to build the english docs without downloading other translations

File: docs/package.json [hunk 2]
```diff
@@ -9,12 +9,20 @@
...
     "deploy": "docusaurus deploy",
+    "write-translations": "docusaurus write-translations",
+    "write-heading-ids": "docusaurus write-heading-ids",
+    "write-i18n": "npm run write-heading-ids && npm run write-translations",
     "format": "prettier --check \"**/*.{js,jsx,json,md,scss}\"",
     "format:fix": "prettier --write \"**/*.{js,jsx,json,md,scss}\"",
     "lint": "set -ex; eslint .",
-    "lint:fix": "npm run lint -- --fix"
+    "lint:fix": "npm run lint -- --fix",
+    "crowdin": "crowdin",
```
Comment (jstarry) [inline docs/package.json:19]: Devs can use `npx crowdin ...` and the scripts can use `crowdin ...` directly, I don't think this shortcut is needed

File: docs/package.json [hunk 3]
```diff
@@ -9,12 +9,20 @@
...
+    "crowdin:download": "npm run write-i18n && npm run crowdin download && npm run format:fix",
+    "crowdin:sync": "npm run write-i18n && npm run crowdin upload && npm run crowdin download && npm run format:fix",
```
Comment (jstarry) [inline docs/package.json:21]: I don't think we need to format files that aren't getting checked into git, and is there a need to write translations before downloading the latest?
```suggestion
    "crowdin:download": "crowdin download",
    "crowdin:upload": "npm run write-i18n && crowdin upload",
```

File: docs/package.json [hunk 4]
```diff
@@ -9,12 +9,20 @@
...
+    "prebuild": "npm run crowdin:download"
   },
   "dependencies": {
+    "@crowdin/cli": "^3.5.2",
```
Comment (jstarry) [inline docs/package.json:25]: Do you mind bumping the below docusaurus versions to `2.0.0-beta.0`?

File: docs/sidebars.js [hunk 5]
```diff
@@ -45,7 +45,6 @@ module.exports = {
       "cli/deploy-a-program",
       "offline-signing",
       "offline-signing/durable-nonce",
-      "cli/usage",
```
Comment (jstarry) [inline docs/sidebars.js:48]: why was this removed?

**Review by jstarry — COMMENTED** [review]
Thanks for the updates @timhagn! The upload should happen in the `build.sh` script before we generate the cli page and update release version numbers. You can look at the `publish-docs.sh` script for how to get the current channel. Uploading should only happen on the stable release* branch.

File: docs/package.json (hunk 3 above)
Comment (timhagn) [reply docs/package.json:21]: Linting broke the build in Travis when the files were not formatted correctly (while they were checked in).
Don't know if Travis will ignore them now when `i18n` is in `.gitignore`.

File: docs/package.json (hunk 1 above)
Comment (timhagn) [reply docs/package.json:22]: Removed it.

File: docs/package.json (hunk 2 above)
Comment (timhagn) [reply docs/package.json:19]: You're probably right, will just add it to the docs.

File: docs/sidebars.js (hunk 5 above)
Comment (timhagn) [reply docs/sidebars.js:48]: This must have fallen through the cracks while merging...

File: docs/package.json (hunk 4 above)
Comment (timhagn) [reply docs/package.json:25]: #done

File: docs/package.json (hunk 3 above)
Comment (timhagn) [reply docs/package.json]: Added your changes, let's try it out in the next master build % ).

File: docs/sidebars.js (hunk 5 above)
Comment (timhagn) [reply docs/sidebars.js]: Re-added it.

File: docs/package.json (hunk 2 above)
Comment (timhagn) [reply docs/package.json]: Added a Translation paragraph to the README.

File: docs/build.sh [hunk 6]
```diff
@@ -8,6 +8,12 @@ source ../ci/env.sh
 
 : "${rust_stable_docker_image:=}" # Pacify shellcheck
 
+# Synchronize translations with Crowdin
+if [[ -n $CI_TAG ]]; then
```
[the remark this answers is not in the document]
Comment (timhagn) [reply docs/build.sh]: I did it according to:

```sh
if [[ -n $CI_TAG ]]; then
  PROJECT_NAME=docs-solana-com
else
...
```

from `publish-docs.sh`,  the switch for other tags afterwards tackles edge & beta.

But replaced it with:

```sh
# Fixated master branch as only one to download translations
ONLY_TRANSLATE_ON='master'
# Grep last name after last possible backslash for current branch
CURRENT_BRANCH=$(git symbolic-ref HEAD | sed -e 's,.*/\(.*\),\1,')

# Synchronize translations with Crowdin only on master branch
if [[ "$CURRENT_BRANCH" = "$ONLY_TRANSLATE_ON" ]]; then
```

[the remark this answers is not in the document]
Comment (timhagn) [reply docs/build.sh]: Dang! Travis checks out commits not branches. Changed it to:

```sh
# Fixate master branch as only one to download translations & get its ref
ONLY_TRANSLATE_ON='master'
MASTER_REF="$(git rev-parse "$ONLY_TRANSLATE_ON")}"
# Get current ref
CURRENT_REF="$(cat ../.git/HEAD)"

# Synchronize translations with Crowdin only on master branch (ref)
if [[ "$CURRENT_REF" = "$MASTER_REF" ]]; then
  echo "Downloading & updating translations..."
  npm run crowdin:download
  npm run crowdin:upload
fi
```

but feeling like the original check might be less wonky % ).
Open for suggestions : )!

[the remark this answers is not in the document]
Comment (jstarry) [reply docs/build.sh]: Hey, I mispoke earlier and we want to make sure that we are only uploading docs on the *stable* branch which you can use the `channel-info.sh` script to determine. I hope that simplifies the issue 😄

[the remark this answers is not in the document]
Comment (jstarry) [reply docs/build.sh]: So the workflow I imagine is that every time we build the stable branch (in this case v1.6) the latest translations will get uploaded to crowdin. By the time we publish docs (usually on the next v1.6.x release) the script will download the latest translations which hopefully by then are up-to-date.

[the remark this answers is not in the document]
Comment (timhagn) [reply docs/build.sh]: Just looked into it & it seems that on my system (Linux Mint 20.1, Kernel 5.4.0-73-generic) `semver.sh`
fails in both zsh & bash, my IDE / shellcheck complains about https://github.com/koalaman/shellcheck/wiki/SC2112 `'function' keyword is non-standard. Delete it.`.

So locally I'm currently unable to check for the stable branch , as the output of the 
`channel-info.sh` script is:

```sh
EDGE_CHANNEL=master
BETA_CHANNEL=v1.6
BETA_CHANNEL_LATEST_TAG=v1.6.9
STABLE_CHANNEL=v1.5
STABLE_CHANNEL_LATEST_TAG=v1.5.19
CHANNEL=
CHANNEL_LATEST_TAG=
```

(or what am I missing oO?) : (

File: docs/build.sh [hunk 7]
```diff
@@ -8,14 +8,27 @@ source ../ci/env.sh
 
 : "${rust_stable_docker_image:=}" # Pacify shellcheck
 
+# Fixate master branch as only one to download translations & get its ref
+ONLY_TRANSLATE_ON='master'
+MASTER_REF="$(git rev-parse "$ONLY_TRANSLATE_ON")}"
```
Comment (t-nelson) [inline docs/build.sh:13]: Stray `}` ?

File: docs/build.sh [hunk 8]
```diff
@@ -8,14 +8,27 @@ source ../ci/env.sh
...
+
 # shellcheck source=ci/rust-version.sh
 source ../ci/rust-version.sh
 ../ci/docker-run.sh "$rust_stable_docker_image" docs/build-cli-usage.sh
 ../ci/docker-run.sh "$rust_stable_docker_image" docs/convert-ascii-to-svg.sh
 ./set-solana-release-tag.sh
 
 # Build from /src into /build
-npm run build
+#npm run build
```
Comment (t-nelson) [inline docs/build.sh:31]: Lets kill the block instead of committing commented code

File: docs/build.sh (hunk 7 above)
Comment (timhagn) [reply docs/build.sh]: Thanks! Fixed.

File: docs/build.sh (hunk 8 above)
Comment (timhagn) [reply docs/build.sh]: Ups, that was a remainder from testing it % ). Uncommented again.

File: docs/build.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (t-nelson) [reply docs/build.sh]: > (or what am I missing oO?) : (

STR?  I don't see `shellcheck` failing in any of the past Buildkite jobs

[the remark this answers is not in the document]
Comment (timhagn) [reply docs/build.sh]: My IDE shellcheck fails with the aforementioned linked error message, not `shellcheck.sh` ; ).
But I still have no clue why `semver.sh` fails...
(By the way, what do you mean with "STR"? "Stands to Reason" is the only thing I could come up with...)

[the remark this answers is not in the document]
Comment (t-nelson) [reply docs/build.sh]: Hmm, maybe just a difference in shellcheck version/config?  CI uses https://github.com/solana-labs/solana/blob/63b97729e6856be4f3955942ac8087c005ed202e/ci/shellcheck.sh

STR is "Steps to reproduce" :sweat_smile:

[the remark this answers is not in the document]
Comment (timhagn) [reply docs/build.sh]: *lol* Sometimes not being a native English speaker makes it hard to understand some acronyms, thanks ^^.
Posted `semver.sh` on https://www.shellcheck.net/ (the site "koalaman" mentions in the [shellcheck repo](https://github.com/koalaman/shellcheck) & it gives me the same errors as my IDE (IntelliJ Studio), e.g.:

```sh
Line 3:
function semverParseInto() {
^-- SC2112: 'function' keyword is non-standard. Delete it.
```
But apart from that I'm still stuck at `channel-info.sh` not returning values for `CHANNEL` & `CHANNEL_LATEST_TAG`.
Should I just do a 

```sh
if [ "$CHANNEL" = "$STABLE_CHANNEL" ]; then
  echo "Downloading & updating translations..."
  npm run crowdin:download
  npm run crowdin:upload
fi

```

in the build script anyhow & be done with it % )? As it seems to work in CI I probably shouldn't mind my workstation - or should we rather look into it?

[the remark this answers is not in the document]
Comment (timhagn) [reply docs/build.sh]: Hey @jstarry & @t-nelson,

any input on how to proceed?

Best,

Tim.

[the remark this answers is not in the document]
Comment (t-nelson) [reply docs/build.sh]: I suspect the `shellcheck` issues you're seeing are from a lack of the `--shell=bash` arg we pass in CI.

`channel-info.sh` is _very_ presumptuous about being run under CI.  I have a branch with a WIP local CI wrapper.  It's not ready to merge though.  I'd recommend hacking values in around [here](https://github.com/solana-labs/solana/blob/b5302e76198b951e7d4f96fbf5f17e273f209dcd/ci/channel-info.sh#L113) and taking care not to merge them

[the remark this answers is not in the document]
Comment (jstarry) [reply docs/build.sh]: all good here @timhagn?

[the remark this answers is not in the document]
Comment (timhagn) [reply docs/build.sh]: Hi  @jstarry & @t-nelson,

wasn't able to do much yesterday, was already off when you replied (only ~2h sleep the night before -.-).

>  I'd recommend hacking values in around here and taking care not to merge them

Now I'm a little confused, which "values" should I hack into your linked `channel-info.sh`, @t-nelson  oO?
As that's identical to the one from `master` & this branch...

Best,

Tim.

[the remark this answers is not in the document]
Comment (t-nelson) [reply docs/build.sh]: I'd recommend making the `build.sh` changes that you suggested in https://github.com/solana-labs/solana/pull/17190#discussion_r634740176, hack a  `CHANNEL` value in to `channel-info.sh` so you can check locally, pull the hack out once it's working as expected, then run it through CI

[the remark this answers is not in the document]
Comment (timhagn) [reply docs/build.sh]: @t-nelson Replaced, adapted & tested it, should work : ).
Now we only have to look into getting the remaining languages working, but AFAIK @rmshea is at it.

[the remark this answers is not in the document]
Comment (jstarry) [reply docs/build.sh]: Great @timhagn! I don't think we need to block on the remaining locales. Any other blockers for this going in?

**Review by t-nelson — DISMISSED** [review]
`*.sh` changes lgtm!

File: docs/build.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (timhagn) [reply docs/build.sh]: @jstarry Nope, should be good to go, LGTM : )!

**Review by jstarry — CHANGES_REQUESTED** [review]
Please remove the zip file and revert the heading id change

**Comment by timhagn** [comment]
> Please remove the zip file and revert the heading id change

Removed the zip file - no clue where that broken one came from, `git blame` didn't show me.
Which ID do you mean?

**Comment by jstarry** [comment]
> > Please remove the zip file and revert the heading id change
> 
> Removed the zip file - no clue where that broken one came from, `git blame` didn't show me.
> Which ID do you mean?

Great! These are the id's I'm referring to: 

Screen Shot 2021-05-25 at 2 37 24 PM

Instead of checking in these id's, they should be generated right before crowdin upload so that us devs don't have to worry about them.

**Comment by timhagn** [comment]
Thnx @jstarry!

Thought you meant an HTML header ^^. Used this regex to find and replace them: `\ \{\#.*`, don't think it broke anything (I hope so, indeed ; ). `crowdin:upload` autogenerates them on build.

Best,

Tim.

**Review by jstarry — APPROVED** [review]
looks great, sorry for the approval delay! Let's give this a shot 🚢

**Review by t-nelson — APPROVED** [review]

**Comment by timhagn** [comment]
Great, LGTM : )!
